uniffi/src/ffi/handle_map.rs

```diff
@@ -0,0 +1,243 @@
+    pub fn delete(&self, h: Handle) -> Result<(), HandleError> {
+        // We use `remove` and not delete (and use the inner block) to ensure
+        // that if `v`'s destructor panics, we aren't holding the write lock
+        // when it happens, so that the map itself doesn't get poisoned.
+        let v = {
+            let mut map = self.map.write().unwrap();
+            map.remove(h)
+        };
+        v.map(drop)
+    }
+
+    /// Convenient wrapper for `delete` which takes a `u64` that it will
+    /// convert to a handle.
+    ///
+    /// The main benefit (besides convenience) of this over the version
+    /// that takes a [`Handle`] is that it allows handling handle-related errors
+    /// in one place.
+    pub fn delete_u64(&self, h: u64) -> Result<(), HandleError> {
+        self.delete(Handle::from_u64(h)?)
+    }
+
+    /// Remove an item from the map, returning either the item,
+    /// or None if its guard mutex got poisoned at some point.
+    ///
+    /// # Locking
+    ///
+    /// Note that this requires taking the map's write lock, and so it will
+    /// block until all other threads have finished any read/write operations.
+    pub fn remove(&self, h: Handle) -> Result<Option<T>, HandleError> {
+        let mut map = self.map.write().unwrap();
+        let v = map.remove(h)?;
+        // TODO sense check.
+        Ok(Some(v))
+    }
+
+    /// Convenient wrapper for `remove` which takes a `u64` that it will
+    /// convert to a handle.
+    ///
+    /// The main benefit (besides convenience) of this over the version
+    /// that takes a [`Handle`] is that it allows handling handle-related errors
+    /// in one place.
+    pub fn remove_u64(&self, h: u64) -> Result<Option<T>, HandleError> {
+        self.remove(Handle::from_u64(h)?)
+    }
+
+    /// Call `callback` with a non-mutable reference to the item from the map,
+    /// after acquiring the necessary locks.
+    ///
+    /// # Locking
+    ///
+    /// Note that this requires taking both:
+    ///
+    /// - The map's read lock, and so it will block until all other threads have
+    ///   finished any write operations.
+    /// - The mutex on the slot the handle is mapped to.
+    ///
+    /// And so it will block if there are ongoing write operations, or if
+    /// another thread is reading from the same handle.
+    ///
+    /// # Panics
+    ///
+    /// This will panic if a previous `get()` or `get_mut()` call has panicked
+    /// inside it's callback. The solution to this
+    ///
+    /// (It may also panic if the handle map detects internal state corruption,
+    /// however this should not happen except for bugs in the handle map code).
+    pub fn get<F, E, R>(&self, h: Handle, callback: F) -> Result<R, E>
+    where
+        F: FnOnce(&T) -> Result<R, E>,
+        E: From<HandleError>,
+    {
+        let map = self.map.read().unwrap();
+        let hm = map.get(h)?;
+        callback(&*hm)
```

Comment:
Hrm, so, this is where we get into the tricky bits. IIUC what we want to do here is:

* Take the read lock
* Read the instance out of the handlemap
* Release the read lock
* Operate on the retrieved instance.

In order to do that, you'll need the lock guard `map` to go out of scope, similar to what's done in the implementation of `delete`. Something like:

```
let obj = {
    let map = self.map.read().unwrap();
    map.get(h)?;
}
callback(&*obj)
```

However! Rust probably won't let that code compile as written above, because it won't let you keep a reference to `&obj` alive after releasing the read lock that was guarding it. If it did, a concurrent thread might remove the object from the handlemap, running its destructor while there's still a live reference held by another thread.

To make the Rust compiler happy, we'll probably need to take a *copy* of the object out of the handlemap, so that it knows other threads can safely drop the instance contained therein:

```
let obj = {
    let map = self.map.read().unwrap();
    map.get(h)?.clone(); // <-- we're going to operate on a copy of the object
}
callback(&*obj)
```

But of course we don't want to actually copy our instances! The semantics we want here are reference counting: don't run the destructor until the instance is removed from the handle *and* there are zero concurrent threads holding a reference to the instance. That's what [Arc](https://doc.rust-lang.org/std/sync/struct.Arc.html) is for.

So, I think we'll have to actually make the type here be:

```
pub struct NonLockingHandleMap<T>
where
    T: Sync + Send,
{
    // Put each instance in an Arc, for refernece-counting.
    pub map: RwLock<HandleMap<Arc<T>>>,
}
```

If we do so, the `.clone()` version of the code above will increment a reference count instead of doing an actual copy, giving us the semantics we want.

Reply:
(This sounds terrifically complicated, but I'm glad of the way Rust is forcing us to work through this explicitly!)

An alternative could be to reach for unsafe code here, since our generated bindings will ensure that we never actually try to drop from the handlemap until all active method calls have completed. But I think I like the additional level of protection against bugs in the generated code, at least for now.

Reply:
I'm also curious if the test as written would have uncovered the fact that this method is unintentionally holding the `RwLock` for the duration of the method call...

Reply:
+1 from me on using an Arc. Even if we control the generated bindings around it right now, we shouldn't pre-optimize that using unsafe code if Arc does the job for us.
IMO it's also good to explicitly call `Arc::clone(&obj)` to make it obvious to the reader that the clone is _cheap_

Reply:
> I'm also curious if the test as written would have uncovered the fact that this method is unintentionally holding the RwLock for the duration of the method call...

I'm not rightly sure.

The test above is calling a long running method on one thread, then calling many times a short running method on another.

Your analysis would suggest the `read` lock around the handle map is preventing concurrent access? So I wonder if the following would happen:

 * the chance of inadvertent locking is low, so if we run the test multiple times, perhaps we'd see a failure?
 * the number of `incrementIfBusy()` calls that increment the count changes between implementations?
 * the test should fail everytime? This is not happening.

Reply:
FWIW: the number of `incrementIfBusy()` calls hasn't noticeably changed between the two implementations.

Reply:
> I'm also curious if the test as written would have uncovered the fact that this method is unintentionally holding the
> RwLock for the duration of the method call

I belatedly realized why the test didn't catch the issue I was worried about here, and wanted to record it for future reference.

This method call was holding a *read* lock for the duration of the call, and since multiple threads can concurrently take a read lock, it would not block running ordinary method calls on other threads. What it would block is any attempt to take a *write* lock on the handlemap, meaning that a long-running method call could have blocked constructors and destructors from running concurrently.

That probably would have been fine for Nimbus, since we only ever construct a single immortal instance. But it still seems worthwhile to use an `Arc` and remove this as a future footgun.